doc/source/testing.rst

```diff
@@ -1,51 +1,124 @@
 Testing
 =======
 
-For local development, it is often desirable to run the scheduler
-process on one's local machine & the workers on minikube.
+During local development, we may want to test the code in three different scenarios, largely
+corresponding to real-world deployment scenarios:
 
-1.  Download, set up and start ``minikube``
-2.  Make it possible for your host to be able to talk to the pods on minikube.
+1. Run code locally (possibly inside a `virtual environment <https://docs.python.org/3/tutorial/venv.html>`_
+2. Dockerize code and run it inside a `docker container <https://www.docker.com/resources/what-container>`_
+3. Run code inside a `Kubernetes Pod <https://kubernetes.io/docs/concepts/workloads/pods/pod/>`_
 
-    On Linux::
+Running *integration tests* with either of these options requires access to a Kubernetes cluster
+to spawn worker pods.
 
-       sudo ip route add 172.17.0.0/16 via $(minikube ip)
+.. note:: In CI, we perform static checks in a container (option 2 above), and run tests in a Pod (option 3 above)
 
-    On OS X::
+Set up local development environment
+------------------------------------
+1. Make sure you have ``make`` installed
+2. Run ``make install`` to install dependencies in your current python environment
 
-       sudo route -n add -net 172.17.0.0/16 $(minikube ip)
+You can now run ``make lint``, ``make test`` etc. - however, many tests will require access
+to a kubernetes cluster, as described below.
 
-    If you get an error message like the following::
+Set up a local Kubernetes cluster
+---------------------------------
+We can set up a kubernetes cluster on the local machine using either
+`minikube <https://minikube.sigs.k8s.io/>`_ or `kind <https://kind.sigs.k8s.io/>`_
 
-       RTNETLINK answers: File exists
+``minikube``
+^^^^^^^^^^^^
+1. Install ``minikube`` by following online instructions, or use::
 
-    it most likely means you have docker running on your host using the same
-    IP range minikube is using. You can fix this by editing your
-    ``/etc/docker/daemon.json`` file to add the following:
+      make kubectl-bootstrap
+2. Start ``minikube`` with::
 
-    .. code-block:: json
+      minikube start
+3. (recommended) to run containerized tests in a pod, make sure you have ``docker`` installed and run ``eval $(shell minikube docker-env)`` before building the Docker image
```

Comment:
Can you link to docker install instructions.

Reply:
:+1: